modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ExecutionServiceImpl.java

```diff
@@ -555,9 +565,16 @@ private QueryPlan prepareFragment(BaseQueryContext ctx, String jsonFragment) {
                             fragment.serialized(),
                             ectx.topologyVersion(),
                             fragmentDesc,
-                            qry.parameters());
+                            fragmentsPerNode.get(nodeId).intValue(),
+                            qry.parameters(),
+                            parametersMarshalled
+                        );
 
                         messageService().send(nodeId, req);
+
+                        // Avoid double marshaling.
```

Comment:
don`t understand comment and avoidance here, can u explain plz?

Reply:
It is quite obvious, just skip marshalling parameters on every iteraration and reuse already done.

Reply:
Marshaled parameters should be sent to each node, instead of marshaling it for each node, we can marshal it once, for the first node and then store it to sent it to other nodes.